designs/_posts/2017-07-13-improved-command-line-reporting.md

```diff
@@ -0,0 +1,540 @@
+do it, it raises questions of ordering and correctness that we will not address 
+here. Bazel will accept 1 value to this flag, for the purpose of helping users 
+understand the origin of the output, not to keep an unbounded log of everything 
+that ever led to an invocation. If additional logging is needed, it might be 
+worth adding state somewhere outside of a flag value.
+
+#### Alternative formats considered
+
++  Passing by command arg
+
+    +  Compiled proto (Format [1])
+
+        +  Simple scripts that want Bazel to log how they were called shouldn't
+           need to write and compile a protobuf message.
+
+    +  Data as a single string (Format [2])
+
+        +  This causes problems with escaping argv to be passed as a 
+           re-combined string. For a complicated command line, this is not 
+           acceptable.
+
+    +  Repeatable arg, each value is an element of argv
+
+        +  We still have potential sub-string escape problems, but now we 
+           introduce dependencies on flag ordering.
+
++  Passing by future structured interface directly over the client-server 
+   grpc interface
+
+    +  This could be a future migration, but we can't block on it for now. It 
+       also would only support tools that depend directly on the interface, 
+       which most will never do, and which none do yet.
+    +  Any future work here would need to ensure that we don't accept both a 
+       command arg and a RunRequest-equivalent argument.
+
+### "Command line" proto for reporting command lines
+
+For each command line described above, it needs a structure. This proposed 
+structure aims to logically group the different parts of the command line, 
+while still being general enough to absorb future changes. This also has the 
+advantage of being flexible enough to be used for passing the "tool" command 
+line, which is not necessarilly structured in the same way as a Bazel command 
+line.
+
+Diagram provided for convenience:
+
+![image](insert_image_url_here)------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

Comment:
I realize I need to actually add the image here. I'll figure out how to upload the image to the website, not sure where these files get stored yet.